pkg/csi_driver/node.go

```diff
@@ -450,3 +477,122 @@ func getFSStat(path string) (available, capacity, used, inodesFree, inodes, inod
 	inodesUsed = inodes - inodesFree
 	return
 }
+
+// nodeStageVolumeUpdateLockInfo updates lock info after NodeStageVolume succeed.
+func (s *nodeServer) nodeStageVolumeUpdateLockInfo(ctx context.Context, req *csi.NodeStageVolumeRequest) error {
+	volumeID := req.GetVolumeId()
+	// No-op if driver does not support lock release.
+	if !s.features.FeatureLockRelease {
+		klog.Infof("Skip lock release info update for volume %s: feature lock release not enabled", volumeID)
+		return nil
+	}
+
+	stagingTargetPath := req.StagingTargetPath
+	klog.Infof("NodeStageVolume succeeded on volume %v to staging target path %s, proceed to lock info configmap updates.", volumeID, stagingTargetPath)
+	// No-op if filestore instance not support lock release.
+	attr := req.GetVolumeContext()
+	if val, ok := attr[attrSupportLockRelease]; !ok || strings.ToLower(val) != "true" {
+		klog.Infof("Lock release is not support on volume %s: missing volume attribute %s. The volume is not dynamically provisioned or the filestore instance is not in enterprise tier", volumeID, attrSupportLockRelease)
+		return nil
+	}
+
+	// Update the configMap after successful nfs mount operation.
+	nodeName := s.driver.config.NodeName
+	configmapName := util.ConfigMapNamePrefix + nodeName
+	klog.Infof("NodeStageVolume getting configmap %s/%s for volume %s", util.ConfigMapNamespace, configmapName, volumeID)
+	cm, err := util.GetConfigMap(ctx, configmapName, util.ConfigMapNamespace, s.kubeClient)
+	if err != nil {
+		klog.Errorf("NodeStageVolume failed to get configmap %s/%s for volume %s: %v", util.ConfigMapNamespace, configmapName, volumeID, err)
+		return err
+	}
+
+	lockInfoKey, err := s.generateLockInfoKeyFromVolumeID(volumeID)
+	if err != nil {
+		klog.Errorf("NodeStageVolume failed to generate lock info key for volume %s: %v", volumeID, err)
+		return err
+	}
+
+	// Create or update the configmap with lock info.
+	filestoreIP := attr[attrIP]
+	if cm == nil {
+		data := map[string]string{lockInfoKey: filestoreIP}
+		klog.Infof("NodeStageVolume creating configmap %s/%s with data %v for volume %s", util.ConfigMapNamespace, configmapName, data, volumeID)
+		cm, err := util.CreateConfigMapWithData(ctx, configmapName, util.ConfigMapNamespace, data, s.kubeClient)
+		if err != nil {
+			klog.Errorf("NodeStageVolume failed to create configmap %s/%s with data %s for volume %s: %v", util.ConfigMapNamespace, configmapName, data, volumeID, err)
+			return err
+		}
+		klog.Infof("NodeStageVolume successfully created configmap %s/%s with data %v for volume %s", cm.Name, cm.Name, cm.Data, volumeID)
+		return nil
+	}
+
+	_, err = util.UpdateConfigMapWithKeyValue(ctx, cm, lockInfoKey, filestoreIP, s.kubeClient)
```

Comment:
does it handle the no-op case.

i.e configMap was updated in the last retry, but for some reason NodeStage failed. Kubelet retries node stage, csi finds staging path already mounted, and configMap retries the same update

i.e not sure what the error code would be here. would be nil if we update configMap with same data?

Reply:
resolving offline